### PR TITLE
JENA-1881: RDF*

### PR DESCRIPTION
**Pull request by afs**
[JENA-1881](https://issues.apache.org/jira/browse/JENA-1881)

* In-memory
* Turtle, SPARQL
* NTriples, NQuads, TriG
* Not RDF/XML
* JSON results, not SPARQL XML results.

### CODE REVIEW

File: jena-arq/src/main/java/org/apache/jena/sparql/function/library/triple/FunctionTripleTerm.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.jena.sparql.expr.NodeValue;
+import org.apache.jena.sparql.function.FunctionBase1;
+
+/** Base class for one argument functions working on a triple term argument. */
+public abstract class FunctionTripleTerm extends FunctionBase1 {
+    public FunctionTripleTerm() { }
+
+    @Override
+    public NodeValue exec(NodeValue nv) {
+        System.out.println(nv);
```
Comment (kinow) [inline jena-arq/src/main/java/org/apache/jena/sparql/function/library/triple/FunctionTripleTerm.java:34]: Debug left-over?

File: jena-arq/src/main/java/org/apache/jena/riot/writer/TurtleShell.java [hunk 2]
```diff
@@ -335,25 +333,39 @@ private boolean isDefaultGraph(Node node) {
...
+            Set<Node> blankNodesInTripleTerms = new HashSet<>();
+            // *** Look inside Triple terms
             ExtendedIterator<Triple> iter = find(ANY, ANY, ANY) ;
             try {
                 for ( ; iter.hasNext() ; ) {
                     Triple t = iter.next() ;
                     Node subj = t.getSubject() ;
                     Node obj = t.getObject() ;
-
+
```
Comment (kinow) [inline jena-arq/src/main/java/org/apache/jena/riot/writer/TurtleShell.java:353]: Extra spaces

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterAddTripleTerm.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.jena.sparql.engine.ExecutionContext;
+import org.apache.jena.sparql.engine.QueryIterator;
+import org.apache.jena.sparql.engine.binding.Binding;
+import org.apache.jena.sparql.engine.binding.BindingFactory;
+import org.apache.jena.sparql.serializer.SerializationContext;
+
+/**
+ * Add {@code (var, matchTriple)} to a {@link QueryIterator}.
+ * <p>
+ * The supplied triple is a triple pattern which is grounded by replacing avriables
```
Comment (kinow) [inline jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterAddTripleTerm.java:36]: s/avriables/variables

File: jena-arq/src/main/java/org/apache/jena/query/ARQ.java [hunk 4]
```diff
@@ -460,7 +460,7 @@ public static void enableOptimizer(Context context, boolean state)
     public static final Symbol optMergeBGPs = SystemARQ.allocSymbol("optMergeBGPs");
 
     /**
-     * Context key controlling whether the standard optimizater applies the optimization
```
Comment (kinow) [inline jena-arq/src/main/java/org/apache/jena/query/ARQ.java:463]: Though a typo, that sounds like a cool name for some project, or sci-fi weapon :rocket:

File: jena-arq/Grammar/sse/tokens.inc [hunk 5]
```diff
@@ -147,12 +146,12 @@ TOKEN:
 TOKEN:
 {
   // Anything left that isn't structural
-  // LPAREN and RPAREN / LBRACKET/RBRACKET
+  // LPAREN and RPAREN / LBRACKET/RBRACKET / LBRACE/BRCAE
```
Comment (kinow) [inline jena-arq/Grammar/sse/tokens.inc:149]: s/`BRCAE`/`RBRACE` ?

File: jena-core/src/main/java/org/apache/jena/graph/JenaNodeException.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+package org.apache.jena.graph;
+
+import org.apache.jena.shared.JenaException;
+
+public class JenaNodeException extends JenaException {
+    public JenaNodeException(String message) {
+        super(message);
+    }
+}
```
[the remark this answers is not in the document]
Comment (kinow) [reply jena-core/src/main/java/org/apache/jena/graph/JenaNodeException.java:27]: (not all, saw some new files that had the newline, but again, that's not really important)

File: jena-arq/src/main/java/org/apache/jena/sparql/function/library/triple/FunctionTripleTerm.java (hunk 1 above)
Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/sparql/function/library/triple/FunctionTripleTerm.java:34]: Oops. It also shows there are no tests for this new functions.  I'll add some.

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterAddTripleTerm.java (hunk 3 above)
Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterAddTripleTerm.java:36]: Done

File: jena-arq/Grammar/sse/tokens.inc (hunk 5 above)
Comment (afs) [reply jena-arq/Grammar/sse/tokens.inc:149]: Done.

File: jena-arq/src/main/java/org/apache/jena/riot/writer/TurtleShell.java (hunk 2 above)
Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/riot/writer/TurtleShell.java:353]: Gone!

File: jena-core/src/main/java/org/apache/jena/graph/JenaNodeException.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (afs) [reply jena-core/src/main/java/org/apache/jena/graph/JenaNodeException.java:27]: My Eclipse template for "new java file" didn't have a newline on the end. Fixed. And fixed new java files.

**Comment by kinow** [comment]
Running Fuseki on this branch now. @afs , my SPARQL-Fu is quite rusty, so could you help me with a simple test query?

I'm using the `books.ttl` shipped with Jena source code. After reading the paper (at least the abstract and intro, without delving into the details), I came up with a query that I think should be use reification.

```sparql
prefix purl: <http://purl.org/dc/elements/1.1/>
prefix vcard: <http://www.w3.org/2001/vcard-rdf/3.0#>
prefix dc: <http://purl.org/dc/elements/1.1/>
prefix : <http://example.org/book/>

SELECT ?authorFamilyName ?bookName
WHERE {
  ?book dc:title ?bookName ;
        dc:creator ?author .
  ?author vcard:N ?authorName .
  ?authorName vcard:Family ?authorFamilyName .          
}
LIMIT 25
```

Then my plan was to replace the reification by RDF*. But Fuseki is failing the following query with "[2020-04-16 15:02:57] Fuseki     WARN  [34] RC = 500 : Attempt to hash something strange: << ?book @dc:creator ?author >>".

```sparql
prefix purl: <http://purl.org/dc/elements/1.1/>
prefix vcard: <http://www.w3.org/2001/vcard-rdf/3.0#>
prefix dc: <http://purl.org/dc/elements/1.1/>
prefix : <http://example.org/book/>

SELECT ?authorFamilyName ?bookName
WHERE {
  ?book dc:title ?bookName .
  <<?book dc:creator ?author>> vcard:N ?authorName .
  ?authorName vcard:Family ?authorFamilyName .          
}
LIMIT 25
```

**Comment by afs** [comment]
First point: `Attempt to hash something strange`. That means it is going to TDB and TDB is not yet supported. Only in-memory datasets. Starting `fuseki --mem /ds` (either full or main).

**Comment by afs** [comment]
Example:
Let's add some triples about triples:

Add this data to books.ttl:
```ttl
<< :book1 dc:creator  "J.K. Rowling" >> dc:source "Bruno" .

<< :book5 dc:creator  "J.K. Rowling" >> dc:source "Andy" .
```
Two triples about triples already in the data.

Then ask a query about sources:

Query:
```sparql
SELECT ?source ?title
WHERE {
  <<?book dc:creator ?author>> dc:source ?source .
  ?book dc:title ?title .
}
```
Run (other window):
```
fuseki --file books.ttl /ds
```
Execute: (query in "Q.arq")
```
s-query --service http://localhost:3030/ds --output text  --file Q.arq
```

**Comment by kinow** [comment]
Worked like a charm!!! Thanks @afs ! Learned a bit about RDF* :rocket:

**Pull request merged**